### DISCUSSION TITLE
How to check if a docker image is already built on GitHub packages

### DISCUSSION THREAD

**Discussion by jactor-rises** [community/community#26999] [Actions]
<p>This question have no answers on stackoverflow and I would appreciate a solution if exists.</p>
<p>Summory of the question:</p>
<p>How can I check if my github workflow allready has build a docker image for this commit?</p>
<p><a href="https://stackoverflow.com/questions/60892354/how-to-check-if-a-version-of-a-github-docker-image-already-exists" rel="nofollow noopener">https://stackoverflow.com/questions/60892354/how-to-check-if-a-version-of-a-github-docker-image-already-exists</a></p>

**Comment by Yanjingzhu** [comment]
<p>Hi <a class="mention" href="/jactor-rises">@jactor-rises</a> ,</p>
<p>You could use <a href="https://docs.docker.com/registry/spec/api/#listing-image-tags" rel="nofollow noopener">Docker Registry</a> API v2 to list image tags . Then you could compare the tags with the current commit sha. </p>
<p>The API url for GitHub Package Registry is like: </p>
<p><a href="https://docker.pkg.github.com/v2/%7Borg%7D/%7Brepo%7D/%7Bimage_name%7D/tags/list" rel="nofollow noopener">https://docker.pkg.github.com/v2/{org}/{repo}/{image_name}/tags/list</a></p>

**Comment by cerna** [comment]
<p>Hi,</p>
<p>you can use the Github Packages preview GraphQL API with connection with <em>bash</em> invocation of <em>jq</em>. <a href="https://github.com/machinekit/machinekit-hal/blob/master/.github/workflows/debian-builder-workflow.yaml#L136-L149" rel="nofollow noopener">This</a> is example of how I am doing it.</p>
<pre><code>query:
   'query getDockerImages(\$owner: String!, \$repository: String!) {
    repository(owner: \$owner, name: \$repository) {
     registryPackages(packageType: DOCKER, publicOnly: true, first: 100) {
      nodes {
       name
       version(version: \"latest\") {
         sha256
       }
     }
   }
  }
}
</code></pre>
<p>And then something like:</p>
<pre><code>test_array=($(printf "$MATRIX_JSON" | jq -r '.include[] | [.architecture, .osVersionNumber|tostring] | join("_")'))
MISSING=0
for i in ${test_array[@]}
do
 IMAGENAME="${IMAGE_NAME_ROOT}$i"
 IMAGESHA=$(jq -r --arg IMAGENAME "$IMAGENAME" '.data.repository.registryPackages.nodes[] | select(.name == $IMAGENAME).version.sha256 | select(.!=null)' ${INPUT_JSON_FILE})
 if [-z "$IMAGESHA"]; then
  printf "Docker image $IMAGENAME:latest does not exist in registry docker.pkg.github.com/$REPOSITORY_FULL\n"
  ((MISSING=MISSING+1))
  fi
done
if [$MISSING -gt 0]; then
  printf "Registry is missing $MISSING Docker image(s)\n"
  echo ::set-env name=BuildDockerImage::true
else
  printf "All images present in registry, no missing packages waiting to be build\n"
fi
</code></pre>

**Comment by jactor-rises** [comment]
<p>The suggested solutions seems to include lots of coding and such code is hard to maintain. From <a href="https://docs.docker.com/registry/spec/api/" rel="nofollow noopener">https://docs.docker.com/registry/spec/api/</a>:</p>
<p><strong>Existing Manifests</strong></p>
<p>The image manifest can be checked for existence with the following url:</p>
<pre><code>&lt;font color="#999999"&gt;HEAD /v2/&amp;lt;name&amp;gt;/manifests/&amp;lt;reference&amp;gt;&lt;/font&gt;
</code></pre>
<p>The name and reference parameter identify the image and are required. The reference may include a tag or digest.</p>
<p>A 404 Not Found response will be returned if the image is unknown to the registry. If the image exists and the response is successful the response will be as follows:</p>
<pre><code>&lt;font color="#999999"&gt;200 OK
Content-Length: &amp;lt;length of manifest&amp;gt;
Docker-Content-Digest: &amp;lt;digest&amp;gt;&lt;/font&gt;
</code></pre>
<p>Would it not be easier to try and get the manifest for a docker image and see if it exists or not? Examples?</p>

**Comment by Yanjingzhu** [comment, marked as answer]
<p>Hi <a class="mention" href="/jactor-rises">@jactor-rises</a> , </p>
<p>You could use docker registry api to get image manifest for Github Package Registry. </p>
<pre><code>on: push

jobs:
  build:
    runs-on: ubuntu-latest
    steps:
    - name: Check package version
      run: |
        manifest=$(curl -X GET https://docker.pkg.github.com/v2/{org}/{repo}/{image name}/manifests/$GITHUB_SHA -u $GITHUB_ACTOR:${{ secrets.GITHUB_TOKEN }} | jq '.')
        echo $manifest
</code></pre>

**Comment by jactor-rises** [comment]
<p><a class="mention" href="/yanjingzhu">@yanjingzhu</a> : do you need some special permissions to do this? From output from use of action <a href="https://github.com/navikt/bidrag-docker/tree/master/exists" rel="nofollow noopener">https://github.com/navikt/bidrag-docker/tree/master/exists</a> (see exists.sh):</p>
<p>1s</p>
<p><a href="https://github.com/navikt/bidrag-dokument/pull/241/checks?check_run_id=658617394#step:2:1" rel="nofollow noopener">1</a> Run navikt/bidrag-docker/exists@v1</p>
<p><a href="https://github.com/navikt/bidrag-dokument/pull/241/checks?check_run_id=658617394#step:2:2" rel="nofollow noopener">2</a> with:</p>
<p><a href="https://github.com/navikt/bidrag-dokument/pull/241/checks?check_run_id=658617394#step:2:3" rel="nofollow noopener">3</a> image_name: bidrag-dokument</p>
<p><a href="https://github.com/navikt/bidrag-dokument/pull/241/checks?check_run_id=658617394#step:2:4" rel="nofollow noopener">4</a> GITHUB_TOKEN: ***</p>
<p><a href="https://github.com/navikt/bidrag-dokument/pull/241/checks?check_run_id=658617394#step:2:5" rel="nofollow noopener">5</a> env:</p>
<p><a href="https://github.com/navikt/bidrag-dokument/pull/241/checks?check_run_id=658617394#step:2:6" rel="nofollow noopener">6</a> IMAGE: <a href="http://docker.pkg.github.com/navikt/bidrag-dokument/bidrag-dokument:d4cc1bcd2ff78387b19db6f15261ba5207beab1c" class="inline-onebox-loading" rel="nofollow noopener">docker.pkg.github.com/navikt/bidrag-dokument/bidrag-dokument:d4cc1bcd2ff78387b19db6f15261ba5207beab1c</a></p>
<p><a href="https://github.com/navikt/bidrag-dokument/pull/241/checks?check_run_id=658617394#step:2:7" rel="nofollow noopener">7</a>/bin/bash /home/runner/work/_actions/navikt/bidrag-docker/v1/exists/dist/…/exists.sh bidrag-dokument</p>
<p><a href="https://github.com/navikt/bidrag-dokument/pull/241/checks?check_run_id=658617394#step:2:8" rel="nofollow noopener">8</a> % Total % Received % Xferd Average Speed Time Time Time Current</p>
<p><a href="https://github.com/navikt/bidrag-dokument/pull/241/checks?check_run_id=658617394#step:2:9" rel="nofollow noopener">9</a> Dload Upload Total Spent Left Speed</p>
<p><a href="https://github.com/navikt/bidrag-dokument/pull/241/checks?check_run_id=658617394#step:2:10" rel="nofollow noopener">10</a></p>
<p><a href="https://github.com/navikt/bidrag-dokument/pull/241/checks?check_run_id=658617394#step:2:11" rel="nofollow noopener">11</a> 0 0 0 0 0 0 0 0 --:–:-- --:–:-- --:–:-- 0</p>
<p><a href="https://github.com/navikt/bidrag-dokument/pull/241/checks?check_run_id=658617394#step:2:12" rel="nofollow noopener">12</a> 0 205 0 0 0 0 0 0 --:–:-- --:–:-- --:–:-- 0</p>
<p><a href="https://github.com/navikt/bidrag-dokument/pull/241/checks?check_run_id=658617394#step:2:13" rel="nofollow noopener">13</a></p>
<p><a href="https://github.com/navikt/bidrag-dokument/pull/241/checks?check_run_id=658617394#step:2:14" rel="nofollow noopener">14</a>Headere i respons fra <a href="https://docker.pkg.github.com/v2/navikt/bidrag-dokument/bidrag-dokument/manifests/d4cc1bcd2ff78387b19db6f15261ba5207beab1c" rel="nofollow noopener">https://docker.pkg.github.com/v2/navikt/bidrag-dokument/bidrag-dokument/manifests/d4cc1bcd2ff78387b19db6f15261ba5207beab1c</a>:</p>
<p><a href="https://github.com/navikt/bidrag-dokument/pull/241/checks?check_run_id=658617394#step:2:15" rel="nofollow noopener">15</a>----</p>
<p><a href="https://github.com/navikt/bidrag-dokument/pull/241/checks?check_run_id=658617394#step:2:16" rel="nofollow noopener">16</a>HTTP/1.1 401 Unauthorized</p>
<p><a href="https://github.com/navikt/bidrag-dokument/pull/241/checks?check_run_id=658617394#step:2:17" rel="nofollow noopener">17</a>Content-Security-Policy: default-src ‘none’;</p>
<p><a href="https://github.com/navikt/bidrag-dokument/pull/241/checks?check_run_id=658617394#step:2:18" rel="nofollow noopener">18</a>Docker-Distribution-Api-Version: registry/2.0</p>
<p><a href="https://github.com/navikt/bidrag-dokument/pull/241/checks?check_run_id=658617394#step:2:19" rel="nofollow noopener">19</a>Server: GitHub Registry</p>
<p><a href="https://github.com/navikt/bidrag-dokument/pull/241/checks?check_run_id=658617394#step:2:20" rel="nofollow noopener">20</a>Strict-Transport-Security: max-age=31536000;</p>
<p><a href="https://github.com/navikt/bidrag-dokument/pull/241/checks?check_run_id=658617394#step:2:21" rel="nofollow noopener">21</a>X-Content-Type-Options: nosniff</p>
<p><a href="https://github.com/navikt/bidrag-dokument/pull/241/checks?check_run_id=658617394#step:2:22" rel="nofollow noopener">22</a>X-Frame-Options: DENY</p>
<p><a href="https://github.com/navikt/bidrag-dokument/pull/241/checks?check_run_id=658617394#step:2:23" rel="nofollow noopener">23</a>X-Xss-Protection: 1; mode=block</p>
<p><a href="https://github.com/navikt/bidrag-dokument/pull/241/checks?check_run_id=658617394#step:2:24" rel="nofollow noopener">24</a>Date: Sat, 09 May 2020 08:06:57 GMT</p>
<p><a href="https://github.com/navikt/bidrag-dokument/pull/241/checks?check_run_id=658617394#step:2:25" rel="nofollow noopener">25</a>Content-Length: 205</p>
<p><a href="https://github.com/navikt/bidrag-dokument/pull/241/checks?check_run_id=658617394#step:2:26" rel="nofollow noopener">26</a>Content-Type: text/plain; charset=utf-8</p>
<p><a href="https://github.com/navikt/bidrag-dokument/pull/241/checks?check_run_id=658617394#step:2:27" rel="nofollow noopener">27</a>X-GitHub-Request-Id: 0400:2CD7:6D029:ECA4F:5EB664A1</p>
<p><a href="https://github.com/navikt/bidrag-dokument/pull/241/checks?check_run_id=658617394#step:2:28" rel="nofollow noopener">28</a></p>
<p><a href="https://github.com/navikt/bidrag-dokument/pull/241/checks?check_run_id=658617394#step:2:29" rel="nofollow noopener">29</a></p>
<p><a href="https://github.com/navikt/bidrag-dokument/pull/241/checks?check_run_id=658617394#step:2:30" rel="nofollow noopener">30</a>Found 0 mentions of HTTP/1.1\ 404</p>
<p><a href="https://github.com/navikt/bidrag-dokument/pull/241/checks?check_run_id=658617394#step:2:31" rel="nofollow noopener">31</a>unauthorized, assuming not found</p>

**Comment by Yanjingzhu** [comment]
<p>When execute curl request, -u  parameter is used to pass user and password. Please make sure the GITHUB_TOKEN is passed to scripts in exists.sh.  </p>
